Fix substraction with overflow in `wrong_number_of_generic_args.rs`

fuzzypixelz opened the pull request:
Fixes #104287

This issue happens in the `suggest_moving_args_from_assoc_fn_to_trait_for_qualified_path` function, which seems to run before the error checking facilities can catch an invalid use of generic arguments. Thus we get a subtraction with overflow because the code implicitly assumes that the source program makes sense (or is this assumption not true even if the program is correct?).

compiler-errors commented on compiler/rustc_hir_analysis/src/structured_errors/wrong_number_of_generic_args.rs:
```diff
@@ -728,7 +728,10 @@ impl<'a, 'tcx> WrongNumberOfGenericArgs<'a, 'tcx> {
             let num_generic_args_supplied_to_trait = trait_path_segment.args().num_generic_params();
 
-            if num_assoc_fn_excess_args == num_trait_generics_except_self - num_generic_args_supplied_to_trait {
+            if num_trait_generics_except_self
+                .checked_sub(num_generic_args_supplied_to_trait)
+                .is_some_and(|diff| diff == num_assoc_fn_excess_args)
```
Is it possible to just do this transformation: `a = b - c` <=> `a + c = b`?

```suggestion
            if num_generic_args_supplied_to_trait + num_assoc_fn_excess_args == num_trait_generics_except_self
```

lcnr commented:
r? @compiler-errors

fuzzypixelz commented:
Should I squash the commits?

compiler-errors commented:
Yes please.

compiler-errors commented:
Also this needs a test

fuzzypixelz commented:
@compiler-errors Sure, I couldn't figure out how to do that kind of testing. Cool guide.

I'm on it.

compiler-errors commented:
Use `@rustbot ready` to mark this as ready for review when you've added a UI test.

fuzzypixelz committed:
Fix substraction with overflow in `wrong_number_of_generic_args.rs`

Rarranging the substration and equality check into an addition and an equality
check is sufficient.

Algebra is cool, isn't it?

Co-authored-by: Michael Goulet <michael@errs.io>

fuzzypixelz committed:
Add a UI test to ensure rustc doesn't do arithmetic overflows

This relies on the CI testing a rustc that's compiled with overflow-checks = true

fuzzypixelz commented:
@rustbot ready

compiler-errors commented:
@bors r+ rollup

bors commented:
:pushpin: Commit 3046af0cf652a4ccecd823d7b5cf81fbf026ab38 has been approved by `compiler-errors`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=compiler-errors 3046af0cf652a4ccecd823d7b5cf81fbf026ab38 -->
<!-- homu: {"type":"Approved","sha":"3046af0cf652a4ccecd823d7b5cf81fbf026ab38","approver":"compiler-errors","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
